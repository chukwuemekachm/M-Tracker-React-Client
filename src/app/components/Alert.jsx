import React from 'react';
import PropTypes from 'prop-types';

const Alert = ({ message, backgroundColor }) => (
  <div className="ch-alert" style={{ backgroundColor }}>
    <p>
      {message}
    </p>
  </div>
);

Alert.propTypes = {
  message: PropTypes.string.isRequired,
  backgroundColor: PropTypes.string.isRequired,
};

export default Alert;
