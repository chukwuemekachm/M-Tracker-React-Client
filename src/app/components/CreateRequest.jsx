import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Alert from './Alert';
import loadingSvg from '../assets/svg/loading.svg';

class CreateRequest extends Component {
  constructor(props) {
    super(props);
    this.state = {
      title: '',
      type: 'none',
      description: '',
    };
    this.onChange = this.onChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
  }

  onChange(event) {
    this.setState({
      [event.target.name]: event.target.value,
    });
  }

  onSubmit(event) {
    event.preventDefault();
    const {
      create, history, toggleCreateModal, getRequests,
    } = this.props;
    create(this.state).then((response) => {
      if (response.code) {
        switch (response.code) {
          case 201:
            toggleCreateModal();
            getRequests();
            history.push('/dashboard');
            break;
          case 401:
            toggleCreateModal();
            history.push('/login');
            break;
          default:
            break;
        }
      }
    });
  }

  render() {
    const { title, type, description } = this.state;
    const {
      toggleCreateModal, error, loading, message,
    } = this.props;
    return (
      <div className="ch-modal">
        <div className="ch-modal-section">
          <div className="ch-col-12 ch-modal-heading">
            <h4>
              New Request
            </h4>
            <span onClick={() => toggleCreateModal()}>
              &times;
            </span>
          </div>
          <div className="ch-row">
            <div className="ch-col-10">
              <div className="ch-row">
                <div className="ch-col-12">
                  {
                    error
                      ? (
                        <Alert
                          message={message}
                          backgroundColor="#E74C3C"
                        />
                      )
                      : ''
                  }
                  {
                    loading
                      ? <img src={loadingSvg} alt="loading" />
                      : ''
                  }
                </div>
              </div>
              <form id="form-create-request" onSubmit={this.onSubmit}>
                <label htmlFor="title" className="ch-label">
                  Title
                </label>
                <input
                  type="text"
                  name="title"
                  id="title"
                  className="ch-input"
                  required
                  onChange={this.onChange}
                  value={title}
                />
                <label htmlFor="type" className="ch-label">
                  Type
                </label>
                <select name="type" id="type" className="ch-input" value={type} onChange={this.onChange}>
                  <option value="none">
                    -- select --
                  </option>
                  <option value="repair">
                    Repair
                  </option>
                  <option value="maintenance">
                    Maintenance
                  </option>
                </select>
                <label htmlFor="description" className="ch-label">
                  Description
                </label>
                <textarea
                  name="description"
                  id="description"
                  className="ch-input ch-input-desc"
                  value={description}
                  required
                  onChange={this.onChange}
                />
                <button className="ch-btn-round" type="submit">
                  Submit Request
                </button>
              </form>
              <br />
            </div>
          </div>
        </div>
      </div>
    );
  }
}


CreateRequest.propTypes = {
  toggleCreateModal: PropTypes.func.isRequired,
  create: PropTypes.func.isRequired,
  history: PropTypes.shape({}).isRequired,
  message: PropTypes.string.isRequired,
  error: PropTypes.bool.isRequired,
  loading: PropTypes.bool.isRequired,
  getRequests: PropTypes.func.isRequired,
};

export default CreateRequest;
