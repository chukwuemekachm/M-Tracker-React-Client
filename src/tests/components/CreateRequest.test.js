import React from 'react';
import { shallow } from 'enzyme';
import CreateRequest from '../../app/components/CreateRequest';
import Alert from '../../app/components/Alert';

const createMock = jest.fn();
createMock.mockResolvedValueOnce({
  code: 201,
});

const props = {
  toggleCreateModal: jest.fn(),
  create: createMock,
  error: true,
  loading: true,
  message: 'No message',
  history: {
    push: route => jest.fn(route),
  },
  getRequests: jest.fn(),
};

describe('Create Request component', () => {
  it('should render without errors', () => {
    const wrapper = shallow(<CreateRequest {...props} />);
    expect(wrapper.find('div')).toHaveLength(7);
    expect(wrapper.find('h4')).toHaveLength(1);
    expect(wrapper.find('span')).toHaveLength(1);
    expect(wrapper.find('img')).toHaveLength(1);
    expect(wrapper.find(Alert)).toHaveLength(1);
    expect(wrapper.find('button')).toHaveLength(1);
    expect(wrapper.find('form')).toHaveLength(1);
    expect(wrapper.find('label')).toHaveLength(3);
    expect(wrapper.find('textarea')).toHaveLength(1);
    expect(wrapper.find('select')).toHaveLength(1);
    expect(wrapper.find('option')).toHaveLength(3);
    expect(wrapper).toMatchSnapshot();
    wrapper.find('[name="title"]').simulate('change', {
      target: { value: 'My fan is burnt' },
    });
    wrapper.find('[name="type"]').simulate('change', {
      target: { value: 'repair' },
    });
    wrapper.find('[name="description"]').simulate('change', {
      target: { value: 'My fan got burnt due to a power surge' },
    });
    wrapper.find('form').simulate('submit', { preventDefault: jest.fn() });
  });
});
