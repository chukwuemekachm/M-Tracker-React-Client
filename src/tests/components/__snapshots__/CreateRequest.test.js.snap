// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Request component should render without errors 1`] = `
<div
  className="ch-modal"
>
  <div
    className="ch-modal-section"
  >
    <div
      className="ch-col-12 ch-modal-heading"
    >
      <h4>
        New Request
      </h4>
      <span
        onClick={[Function]}
      >
        Ã—
      </span>
    </div>
    <div
      className="ch-row"
    >
      <div
        className="ch-col-10"
      >
        <div
          className="ch-row"
        >
          <div
            className="ch-col-12"
          >
            <Alert
              backgroundColor="#E74C3C"
              message="No message"
            />
            <img
              alt="loading"
              src={null}
            />
          </div>
        </div>
        <form
          id="form-create-request"
          onSubmit={[Function]}
        >
          <label
            className="ch-label"
            htmlFor="title"
          >
            Title
          </label>
          <input
            className="ch-input"
            id="title"
            name="title"
            onChange={[Function]}
            required={true}
            type="text"
            value=""
          />
          <label
            className="ch-label"
            htmlFor="type"
          >
            Type
          </label>
          <select
            className="ch-input"
            id="type"
            name="type"
            onChange={[Function]}
            value="none"
          >
            <option
              value="none"
            >
              -- select --
            </option>
            <option
              value="repair"
            >
              Repair
            </option>
            <option
              value="maintenance"
            >
              Maintenance
            </option>
          </select>
          <label
            className="ch-label"
            htmlFor="description"
          >
            Description
          </label>
          <textarea
            className="ch-input ch-input-desc"
            id="description"
            name="description"
            onChange={[Function]}
            required={true}
            value=""
          />
          <button
            className="ch-btn-round"
            type="submit"
          >
            Submit Request
          </button>
        </form>
        <br />
      </div>
    </div>
  </div>
</div>
`;
